File: aicode/minicli4/typescript/src/markdown.ts
import { ansi } from "./ansi";

export interface MarkdownRenderOptions {
  colorize?: boolean;
}

function style(text: string, fn: (value: string) => string, colorize: boolean): string {
  return colorize ? fn(text) : text;
}

function renderInline(line: string, colorize: boolean): string {
  const codeTokens: string[] = [];
  let out = line.replace(/`([^`]+)`/g, (_m, code: string) => {
    const token = `\u0000CODE${codeTokens.length}\u0000`;
    codeTokens.push(style(code, ansi.code, colorize));
    return token;
  });
  out = out.replace(/\[([^\]]+)\]\(([^)]+)\)/g, (_m, label: string, url: string) => {
    const decoratedLabel = style(label, ansi.cyan, colorize);
    const decoratedUrl = style(` (${url})`, ansi.gray, colorize);
    return `${decoratedLabel}${decoratedUrl}`;
  });
  out = out.replace(/\*\*([^*]+)\*\*/g, (_m, text: string) => style(text, ansi.bold, colorize));
  out = out.replace(/\*([^*]+)\*/g, (_m, text: string) => style(text, ansi.italic, colorize));
  out = out.replace(/^#{1,6}\s+/, "");
  out = out.replace(/\u0000CODE(\d+)\u0000/g, (_m, idx: string) => codeTokens[Number(idx)] ?? "");
  return out;
}

function renderLine(line: string, state: { inFence: boolean }, colorize: boolean): string {
  if (/^\s*```/.test(line)) {
    state.inFence = !state.inFence;
    return style(state.inFence ? "┌ code" : "└ end code", ansi.magenta, colorize);
  }

  if (state.inFence) {
    return style(line, ansi.code, colorize);
  }

  if (/^#{1,6}\s+/.test(line)) {
    return style(renderInline(line, colorize), ansi.bold, colorize);
  }
  if (/^>\s?/.test(line)) {
    return style(`| ${renderInline(line.replace(/^>\s?/, ""), colorize)}`, ansi.gray, colorize);
  }
  if (/^\s*([-*+]|\d+\.)\s+/.test(line)) {
    const content = line.replace(/^\s*([-*+]|\d+\.)\s+/, "");
    return `${style("•", ansi.blue, colorize)} ${renderInline(content, colorize)}`;
  }
  return renderInline(line, colorize);
}

export class MarkdownAnsiStreamRenderer {
  private remainder = "";
  private readonly state = { inFence: false };
  constructor(private readonly options: MarkdownRenderOptions = {}) {}

  write(chunk: string): string {
    this.remainder += chunk;
    let output = "";
    let newlineAt = this.remainder.indexOf("\n");
    while (newlineAt >= 0) {
      const line = this.remainder.slice(0, newlineAt);
      this.remainder = this.remainder.slice(newlineAt + 1);
      output += `${renderLine(line, this.state, Boolean(this.options.colorize))}\n`;
      newlineAt = this.remainder.indexOf("\n");
    }
    return output;
  }

  flush(): string {
    if (!this.remainder) {
      return "";
    }
    const line = this.remainder;
    this.remainder = "";
    return renderLine(line, this.state, Boolean(this.options.colorize));
  }
}

File: aicode/minicli4/python/src/minicli4_py/markdown.py
from __future__ import annotations

import re
from dataclasses import dataclass

from .ansi import ansi


@dataclass(slots=True)
class MarkdownRenderOptions:
    colorize: bool = False


def _style(text: str, fn, colorize: bool) -> str:
    return fn(text) if colorize else text


def _render_inline(line: str, colorize: bool) -> str:
    code_tokens: list[str] = []

    def replace_code(match: re.Match[str]) -> str:
        token = f"\x00CODE{len(code_tokens)}\x00"
        code_tokens.append(_style(match.group(1), ansi.code, colorize))
        return token

    out = re.sub(r"`([^`]+)`", replace_code, line)
    out = re.sub(
        r"\[([^\]]+)\]\(([^)]+)\)",
        lambda m: f"{_style(m.group(1), ansi.cyan, colorize)}{_style(f' <{m.group(2)}>', ansi.gray, colorize)}",
        out,
    )
    out = re.sub(r"\*\*([^*]+)\*\*", lambda m: _style(m.group(1), ansi.bold, colorize), out)
    out = re.sub(r"\*([^*]+)\*", lambda m: _style(m.group(1), ansi.italic, colorize), out)
    out = re.sub(r"^#{1,6}\s+", "", out)
    out = re.sub(r"\x00CODE(\d+)\x00", lambda m: code_tokens[int(m.group(1))] if int(m.group(1)) < len(code_tokens) else "", out)
    return out


class MarkdownAnsiStreamRenderer:
    def __init__(self, options: MarkdownRenderOptions | None = None) -> None:
        self.options = options or MarkdownRenderOptions()
        self.remainder = ""
        self.in_fence = False
        self.fence_lang = ""
        self.code_line = 1

    def _render_line(self, line: str) -> str:
        colorize = self.options.colorize
        fence = re.match(r"^\s*```\s*([a-zA-Z0-9_-]+)?\s*$", line)
        if fence:
            if not self.in_fence:
                self.in_fence = True
                self.fence_lang = fence.group(1) or "text"
                self.code_line = 1
                return _style(f"[code:{self.fence_lang}]", ansi.magenta, colorize)
            self.in_fence = False
            self.fence_lang = ""
            self.code_line = 1
            return _style("[/code]", ansi.magenta, colorize)
        if self.in_fence:
            numbered = f"{self.code_line:>3} | {line}"
            self.code_line += 1
            return _style(numbered, ansi.code, colorize)
        heading = re.match(r"^(#{1,6})\s+(.+)$", line)
        if heading:
            marker = _style(heading.group(1), ansi.yellow, colorize)
            return f"{marker} {_style(_render_inline(heading.group(2), colorize), ansi.bold, colorize)}"
        if re.match(r"^>\s?", line):
            return f"{_style('|', ansi.gray, colorize)} {_render_inline(re.sub(r'^>\s?', '', line), colorize)}"
        ordered = re.match(r"^\s*(\d+)\.\s+(.+)$", line)
        if ordered:
            return f"{_style(f'{ordered.group(1)}.', ansi.blue, colorize)} {_render_inline(ordered.group(2), colorize)}"
        unordered = re.match(r"^\s*[-*+]\s+(.+)$", line)
        if unordered:
            return f"{_style('-', ansi.blue, colorize)} {_render_inline(unordered.group(1), colorize)}"
        table_row = re.match(r"^\s*\|(.+)\|\s*$", line)
        if table_row:
            raw = [cell.strip() for cell in table_row.group(1).split("|")]
            is_div = all(re.match(r"^:?-{3,}:?$", cell) for cell in raw)
            if is_div:
                return _style(f"| {' | '.join(cell.replace(':', '-') for cell in raw)} |", ansi.gray, colorize)
            rendered = [_render_inline(cell, colorize) for cell in raw]
            sep = f" {_style('|', ansi.gray, colorize)} "
            return f"{_style('|', ansi.gray, colorize)} {sep.join(rendered)} {_style('|', ansi.gray, colorize)}"
        return _render_inline(line, colorize)

    def write(self, chunk: str) -> str:
        self.remainder += chunk
        out = ""
        while "\n" in self.remainder:
            line, self.remainder = self.remainder.split("\n", 1)
            out += f"{self._render_line(line)}\n"
        return out

    def flush(self) -> str:
        if not self.remainder:
            return ""
        line = self.remainder
        self.remainder = ""
        return self._render_line(line)
